### PR TITLE
distribute training in fp16

### PR DESCRIPTION
**Pull request by solin319**
Update kvstore, make it can support fp16.
 1. Use copyfromto() to convert fp32 to fp16 before pass the key-value to PSLite.
 2.  In kvstore server, conver fp16 to fp32 and compute in cpu. If we want to compute in fp16, we must pass them to GPU first. Because it's very slow for cpu to compute fp16.
 3. We test several networks in 2 distributed machines. In test, weights and gradients were all in fp16. The speed result is shown below. 'fp16-server-16gpu' means kvstore server compute in fp16 format on gpu. 'fp16-server-32cpu' means kvstore server compute in fp32 format on cpu.

samples/sec

KVStore | dist_sync_device | dist_sync_device | dist_sync_device | dist_sync | dist_sync
-- | -- | -- | -- | -- | --
  | fp32 | fp16-server-16gpu | fp16-server-32cpu | fp16-server-16gpu | fp16-server-32cpu
alexnet | 715.52 | 1300 | - | 326 | -
inception_v3 | 231.82 | 182.23 | - | 238.93 | 243.05
inception_v4 | 112.48 | 90.75 | - | 116.72 | 116.4
resnet_50 | 368.38 | 311.09 | - | 387.77 | 392.24
resnet_101 | 222 | 185.88 | - | 232.47 | 234.51
resnet_152 | 156.82 | 122.3 | - | 165.82 | 166.9
vgg11 | 218.16 | 382.96 | 164.32 | - | -
vgg16 | 197.78 | 292.07 | 181.58 | - | -
vgg19 | 152.53 | 260.44 | 143.16 | - | -

The train accuracy are almost same between training in fp16 and original fp32.

step | fp16 | fp32
-- | -- | --
4550 | 0.421875 | 0.424844
4600 | 0.432812 | 0.434844
4650 | 0.429219 | 0.427812
4700 | 0.426719 | 0.425469
4750 | 0.423594 | 0.43375
4800 | 0.434375 | 0.435781
4850 | 0.442031 | 0.44
4900 | 0.43 | 0.443906
4950 | 0.449688 | 0.449375
5000 | 0.446406 | 0.451562

### CODE REVIEW

**Comment by piiswrong** [comment]
@eric-haibin-lin @rahul003 I think this should be merged with the 2bit PR to make a more general n-bit gradient compression feature.

**Comment by eric-haibin-lin** [comment]
@rahul003 how hard is it to make a general n-bit compression feature?

File: python/mxnet/kvstore_server.py [hunk 1]
```diff
@@ -72,12 +72,12 @@ def run(self):
...
         check_call(_LIB.MXKVStoreRunServer(self.handle, _ctrl_proto(self._controller()), None))
 
-def _init_kvstore_server_module():
+def _init_kvstore_server_module(data_type='float32'):
     """Start server/scheduler."""
     is_worker = ctypes.c_int()
     check_call(_LIB.MXKVStoreIsWorkerNode(ctypes.byref(is_worker)))
     if is_worker.value == 0:
-        kvstore = create('dist')
+        kvstore = create('dist', data_type)
```
Comment (rahul003) [inline python/mxnet/kvstore_server.py]: How do you start the server with float16 dtype?

**Comment by rahul003** [comment]
I agree we should provide similar interfaces for this and 2bit right now, at least on the user facing API first.

A problem is that this solution requires creating a server for this dtype of 16bit. Which means we need to know type of compression when creating a kvstore. In the 2bit PR, since we still use float32 to represent values (for 2bit compression, 16 values represented in one float32), the original server was sent a message to know the type of compression.

The only way I see to merge interfaces is to have compression parameters even in my case to come in at the time of creating kvstore. To start this, I would like to understand how to give the argument defined in python/mxnet/kvstore_server.py to start the server.

General n-bit compression can be done. But neither this PR nor my PR can be directly extended as of now. In my PR, although the framework in kvstores can be generalized easily, the operator is 2bit specific. Either that can be extended by other having operators for different specific n, or if possible coming up with a generic operator.

File: python/mxnet/kvstore_server.py (hunk 1 above)
Comment (solin319) [reply python/mxnet/kvstore_server.py]: In image classification example, we start kvstrore as below.
In file 'common/fit.py'
```
    mx.kvstore_server._init_kvstore_server_module(args.dtype)
    kv = mx.kvstore.create(args.kv_store, args.dtype)
```
In file 'python/mxnet/kvstore_server.py', we delete ‘_init_kvstore_server_module()’ to avoid auto start kvstore server during import mxnet.
```
#_init_kvstore_server_module()
```

**Comment by piiswrong** [comment]
My mistake, I meant general compression interface, not general n-bit compression interface.
For now we can have 2bit and float16 after merging these 2 PRs.

Can't you just send compression='float16' to the server? Similar to how you sent '2bit'

File: src/kvstore/kvstore_dist.h [hunk 2]
```diff
@@ -367,7 +396,7 @@ class KVStoreDist : public KVStoreLocal {
 #if MKL_EXPERIMENTAL == 1
       mkl_set_tblob_eager_mode(send_buf.data());
 #endif
-      real_t* data = send_buf.data().dptr<real_t>();
+      DType* data = send_buf.data().dptr<real_t>();
```
Comment (BiranLi) [inline src/kvstore/kvstore_dist.h]: Why send_buf get dptr data with real_t type when it get dptr data with DType type in line 327?
I got a compiler's error here.

**Comment by rahul003** [comment]
@piiswrong, its a bit tricky because in kvstore.cc, the creation of kvstore_dist_server requires the template argument which specifies in this what dtype to use. 

If we refactor the classes such that there is a class DistServerWrapper (with better name of course) which receives message from worker with dtype and then creates the DistServer, then it can work I think. The wrapper only needs to perform this receiving of commands I think. Everything else can be in dist_server. 

@solin319 I'm not sure if the current way of ```mx.kvstore_server._init_kvstore_server_module(args.dtype)``` is a good interface. What do others think? The above might work unless I missed something.

**Comment by eric-haibin-lin** [comment]
Will fp16 used for all keys? Do we consider the case where we only want to use fp16 for a particular key?

**Comment by solin319** [comment]
Yes, all keys are used in fp16. Because ps_worker_ used in program was defined with template argument. I think it's hard to define two different ps_worker_ to push fp16 and fp32 separately.
@eric-haibin-lin

**Comment by solin319** [comment]
1. In current way, I think the class kvstore_dist is just like class DistServerWrapper mentioned above. We pass the data type to kvstore_dist first, and then create kvstore_dist_server in the function 'RunServer'.
2. I think merge the logic in '_init_kvstore_server_module' to the function 'kvstore.create' may be a better way to start server and worker.

```
from kvstore_server import KVStoreServer

def create(name='local', data_type='float32'):
    if not isinstance(name, string_types):
        raise TypeError('name must be a string')
    handle = KVStoreHandle()
    check_call(_LIB.MXKVStoreCreate(c_str(name), c_str(data_type),
                                    ctypes.byref(handle)))
    kv = KVStore(handle)
    is_worker = ctypes.c_int()
    check_call(_LIB.MXKVStoreIsWorkerNode(ctypes.byref(is_worker)))
    if is_worker.value == 0:
        server = KVStoreServer(kvstore)
        server.run()
        sys.exit()
    return kv
```

File: src/kvstore/kvstore_dist.h [hunk 3]
```diff
@@ -45,12 +45,14 @@ namespace kvstore {
  * it's the server node's job to control the data consistency among all
  * workers. see details on \ref ServerHandle::Start
  */
+
+template <typename DType>
 class KVStoreDist : public KVStoreLocal {
```
Comment (BiranLi) [inline src/kvstore/kvstore_dist.h]: Is function Init also need to specify the dtype?

Comment (solin319) [reply src/kvstore/kvstore_dist.h]: Yes, pslite worker is defined in the init function.
We must let PSLite know which type data to transmit.

Comment (BiranLi) [reply src/kvstore/kvstore_dist.h]: But how to do it? I have checked your code in my mxnet, but it does not work. I have trained epoch 1 in single node with multiprecision and epoch 2 in 4 nodes. Epoch 1 worked well while epoch 2's acc decreased util 0.

Comment (BiranLi) [reply src/kvstore/kvstore_dist.h]: I have created KVstore with 'float16' and inited KVstroe_Server with dtype 'float16' .

Comment (solin319) [reply src/kvstore/kvstore_dist.h]: 1. Try to train epoch 1 and 2 in same environment, such as 4 nodes.
2. How did you start training? I can make a test in my environment.

Comment (BiranLi) [reply src/kvstore/kvstore_dist.h]: Sorry, I have make a mistake in kvstore optimizer codes. I was tortured for several days with this error code. It can work well now.

**Comment by piiswrong** [comment]
Closing due to outdated. @rahul003 please guide @solin319 to merge this into the gradient compression framework.

**Comment by rahul003** [comment]
@solin319 Which machines did you run the above numbers on? Let us try to come up with an easier interface for this so we can use this on the latest Nvidia Gpus. 

Regarding 'I think merge the logic in '_init_kvstore_server_module' to the function 'kvstore.create' may be a better way to start server and worker.': 
But this would mean that the server is created when the code calls kvstore.create(). This has the effect that we end up doing everything in the training script that was written before creation of kvstore. This could possibly allocate memory for data, model, etc. Or we have to instruct users to create the kvserver first (at which point the server process goes into a loop, so other code isn't run), but this seems hacky for an official way of doing things

File: src/kvstore/kvstore_dist_server.h [hunk 4]
```diff
@@ -378,25 +379,46 @@ class KVStoreDistServer {
...
+        recved_tmp = NDArray(dshape, Context::CPU(0), false, mshadow::DataType<real_t>::kFlag);
+        CopyFromTo(recved, &recved_tmp, 0);
+        recved_tmp.WaitToRead();
+      }
       if (stored.is_none()) {
         // initialization
-        stored = NDArray(dshape, Context());
-        CopyFromTo(recved, &stored, 0);
+        if (recved.dtype() != mshadow::DataType<real_t>::kFlag) {
+          stored = NDArray(dshape, Context::CPU(0), false, mshadow::DataType<real_t>::kFlag);
```
Comment (rahul003) [inline src/kvstore/kvstore_dist_server.h]: Why stored received data of fp16 say in the form of fp32?

Comment (solin319) [reply src/kvstore/kvstore_dist_server.h]: Because the compute in fp16 is slow in cpu.
When the size of  received data is large, it's better to move it to gpu with fp16.
When the size of  received data is small, it's better to move it to cpu with fp32.
It's better to implement like below.
```
if (total_size > 100000) {
	          stored = NDArray(dshape, Context::GPU(0), false, mshadow::DataType<DType>::kFlag);
} else{
	          stored = NDArray(dshape, Context::CPU(0), false, mshadow::DataType<real_t>::kFlag);
}
```

File: src/kvstore/kvstore_dist.h [hunk 5]
```diff
@@ -272,16 +282,35 @@ class KVStoreDist : public KVStoreLocal {
...
-            send_buf = NDArray(storage_type, merged.shape(), pinned_ctx_, true, merged.dtype());
+            send_buf = NDArray(storage_type, merged.shape(), pinned_ctx_,
+                               true, mshadow::DataType<DType>::kFlag);
           }
         }
-        CopyFromTo(merged, &send_buf);
+        if (merged.dtype() == mshadow::DataType<DType>::kFlag) {
+          CopyFromTo(merged, &send_buf);
+        } else {
+          NDArray tmp = NDArray(merged.shape(), pinned_ctx_, true, merged.dtype());
```
Comment (rahul003) [inline src/kvstore/kvstore_dist.h]: Here it looks like if the merged is on gpu, then it is copied to CPU. Why two copies? Both tmp and send_buf are on pinned_ctx, right? And if converting to fp16 is faster on GPU then why are we not doing that here?

Comment (solin319) [reply src/kvstore/kvstore_dist.h]: 1. Copy two times, because it doesn't supports to copy from gpu with fp32 to cpu with fp16.  
2. Kvstore_dist just merge grads on all gpus in one machine.
Kvstore_dist_server merge the grads on all the machines.

**Comment by rahul003** [comment]
I have merged and rebased this onto master, and I see this error.

```what():  [15:42:31] src/kvstore/././kvstore_dist_server.h:484: Check failed: req_data.vals.size() == (size_t)req_data.lens[0] (6 vs. 3)```

Trying to debug it. But I don't really understand why req_data.vals.size() becomes 6. On worker's side it is trying to send an array of size 3. Have you seen this ever? 

The code is on https://github.com/rahul003/mxnet/tree/fp16

File: src/kvstore/kvstore_dist_server.h (hunk 4 above)
Comment (rahul003) [reply src/kvstore/kvstore_dist_server.h]: ok thanks

**Comment by rahul003** [comment]
Wondering if this has to do with fp16 being half of fp32, and 6/3=2...

**Comment by rahul003** [comment]
I'm seeing that fp16 with dist_sync_device is slower than fp32 on a cluster of 2 p3.8x instances. (volta gpus)

imagenet-benchmark-alexnet
fp32: 215 samples/sec
fp16: 28 samples/sec

cifar10-resnet
fp32: 760 samples/sec
fp16: 330 samples/sec

**Comment by solin319** [comment]
Witch data type were used in training?
We ues fp16 in training computation.
@rahul003

**Comment by BiranLi** [comment]
@solin319  
Is it possible to consider gradient diffusion in computational calculations, such as adding a grad_scale processing interface?

**Comment by rahul003** [comment]
@solin319 It was being trained in fp16 too. I also tried implementing a system which just converts data to fp32 in kv_dist and sends weights as fp32. That is much faster than this method


@BiranLi could you expand on what that means?

**Comment by BiranLi** [comment]
@rahul003 
Because of the limited range of expression of FP16, gradient diffusion occurs in BP. The easiest way to handle this is to scale the gradient.

**Comment by solin319** [comment]
@rahul003 
For alexnet, try to use fp16 with GPU in kvstore_dist_server.
For resnet, try to use dist_sync.
